oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBZipkinQueryDAO.java

```diff
@@ -0,0 +1,297 @@
+                        if (StringUtil.isNotEmpty(serviceName)) {
+                            query.and(eq(ZipkinServiceRelationTraffic.SERVICE_NAME, serviceName));
+                        }
+                        query.setLimit(QUERY_MAX_SIZE);
+                    }
+                }
+            );
+        final List<String> remoteServices = new ArrayList<>();
+        for (final DataPoint dataPoint : resp.getDataPoints()) {
+            remoteServices.add(dataPoint.getTagValue(ZipkinServiceRelationTraffic.REMOTE_SERVICE_NAME));
+        }
+        return remoteServices;
+    }
+
+    @Override
+    public List<String> getSpanNames(final String serviceName) throws IOException {
+        MeasureQueryResponse resp =
+            query(ZipkinServiceSpanTraffic.INDEX_NAME,
+                  SPAN_TRAFFIC_TAGS,
+                  Collections.emptySet(), new QueryBuilder<MeasureQuery>() {
+
+                    @Override
+                    protected void apply(MeasureQuery query) {
+                        if (StringUtil.isNotEmpty(serviceName)) {
+                            query.and(eq(ZipkinServiceSpanTraffic.SERVICE_NAME, serviceName));
+                        }
+                        query.setLimit(QUERY_MAX_SIZE);
+                    }
+                }
+            );
+        final List<String> spanNames = new ArrayList<>();
+        for (final DataPoint dataPoint : resp.getDataPoints()) {
+            spanNames.add(dataPoint.getTagValue(ZipkinServiceSpanTraffic.SPAN_NAME));
+        }
+        return spanNames;
+    }
+
+    @Override
+    public List<Span> getTrace(final String traceId) throws IOException {
+        StreamQueryResponse resp =
+            query(ZipkinSpanRecord.INDEX_NAME, TRACE_TAGS,
+                  new QueryBuilder<StreamQuery>() {
+
+                      @Override
+                      protected void apply(StreamQuery query) {
+                          query.and(eq(ZipkinSpanRecord.TRACE_ID, traceId));
+                          query.setLimit(QUERY_MAX_SIZE);
+                      }
+                  }
+            );
+
+        List<Span> trace = new ArrayList<>(resp.getElements().size());
+
+        for (final RowEntity rowEntity : resp.getElements()) {
+            ZipkinSpanRecord spanRecord = new ZipkinSpanRecord.Builder().storage2Entity(
+                new BanyanDBConverter.StorageToStream(ZipkinSpanRecord.INDEX_NAME, rowEntity));
+            trace.add(ZipkinSpanRecord.buildSpanFromRecord(spanRecord));
+        }
+
+        return trace;
+    }
+
+    @Override
+    public List<List<Span>> getTraces(final QueryRequest request, Duration duration) throws IOException {
+        final int tracesLimit = request.limit();
+        int scrollLimit = 1000;
+        int scrollFrom = 0;
+        Set<String> traceIds = new HashSet<>();
+        while (traceIds.size() < tracesLimit) {
+            Set<String> resp = getTraceIds(request, duration, scrollFrom, scrollLimit);
```

Comment:
@hanahmily @lujiajing1126 Please confirm this is good enough. 

Zipkin provides trace IDs query to the Span table, and also requires distinct trace IDs according to the `limit` condition. It is using `aggregation` or `distinct .. group by` in other storage options, but BanyanDB doesn't provide this.
So, we have to do a rolling query, but as BanyanDB can't provide a rolling query(no rolling ID concept), we have to use `from... to...`

Reply:
Shall we consider using an extra condition, e.g. `ZipkinSpanRecord.TIMESTAMP_MILLIS` instead of using `limit` and `offset`?
WDYT @hanahmily

Reply:
Time range is user condition, not ours. Unless we use some kind of order by and rolling according to last read record. But is the timestamp accurate enough?
 Is it millisecond? Or it is just second but formated in millisecond format? @wankai123

Reply:
The milliseconds in the span and storage are formatted from nanoseconds.
The milliseconds in the query condition are formatted from seconds, as a user can not select a milliseconds time.